### ISSUE TITLE
Follow-up questions on Bayesian calibration example in the vignette

### ISSUE THREAD

**Issue by GakenouOF** [trotsiuk/r3PG#56]
Dear @florianhartig @trotsiuk 

I tried running the Bayesian calibration using the data provided in vignette build, the result I got was different from your result. I ran both morris sensitivity and Bayesian calibration. The morris sensitivity seems to be the same. I must say here that I don't have advanced knowledge of Bayesian statistics or using BayesianTools in R.   
 
Here is what I observed. 

running this line of  code

`mcmc_out <- runMCMC(`
     `bayesianSetup = mcmc_setup,`
     `sampler = "DEzs",`
     `settings = list(iterations = 4e+03, nrChains = 3))`

On the console, i saw this 

> runMCMC terminated after 37.8599999999999seconds002 . Current logp  -474.9476 -469.4768 -468.8282 . Please wait! 
> runMCMC terminated after 34.8800000000001seconds002 . Current logp  -341.9094 -339.1239 -342.8411 . Please wait! 
> runMCMC terminated after 34.3899999999999seconds002 . Current logp  -357.7484 -358.9804 -356.8762 . Please wait! 

Question: was it suppose to terminate? is it doing the right thing here?

also, the result of `gelmanDiagnostics( mcmc_out)` gave me 95.2, contrary to 1.11 you got. 

I have attached the graph showing the observed, calibrated, and default value. 
![calibration](https://user-images.githubusercontent.com/48384837/124307151-b0d81e00-db67-11eb-96d6-587bfc6d30e5.jpeg)

Please, can you help clarify what i am doing wrong? 

Also, i am doing multiple runs for my simulations, we wrote a kind of pipeline to make r3pg loop over multiple sites. I was trying to go over this method of calibrating the r3pg model and see how i can use it. 

Please, i will also like to get your advice on how to do that since this vignette explains for a single run.

Thank you.

**Comment by trotsiuk** [comment]
Dear @GakenouOF 

Indeed, the code presented in the vignette indicates only `4e+03` iterations (I will correct it in the next commit). But, as described in the vignette text, the calibration run for **4e+06** iteration: `We will run 3 parallel chains, each on a separate computer node for 4e+06 iterations.` Therefore, unfortunately with  `4e+03` in your example the chains didn't converged and you have such a wide range of posterior. Please consider running the MCMC for the longer time.

For multiple sites you can run the log likelihood function for each site and then sum it together.

**Comment by GakenouOF** [comment]
Hi @trotsiuk,

Thank you for your response.  

Regarding the calibration for multiple sites, do you mean i should run the `r3pg_sim` and `r3pg_11` for each site, and then make the summation of `par_cal_best` values for those sites as my `par_cal_best`? 

Please, let me know if i am wrong. 

Thank you!

**Comment by trotsiuk** [comment]
Hi @GakenouOF ,

Not really. What I'm trying to say, is that your `r3pg_ll` function shall calculate the likelihood for all the sites together.

```
r3pg_sim <- function ( par_df, side_id ){
  #' run the simulation for a single selected site, based on the `site_id`
}

r3pg_ll <- function( par_v ){
  #' calculate the likelihood for all the sites together
  
  err_id = grep('err', par_cal_names)
  
  par_df = data.frame(parameter = par_cal_names[-err_id], piab = par_v[-err_id])
  err_v = par_v[err_id]
  
  # logpost for all sites together
  logpost_all_sites <- c()
  
  for( i_site in ALL_YOUR SITES){
    # simulate the model
    sim.df <- r3pg_sim(par_df, site_id = i_site, ....)
    sim.df <- cbind(sim.df[,,2,c(3,5,6)], sim.df[,,4,c(1,2,3)])
    
    # calculate the log likelihood
    logpost <- sapply(1:6, function(i) {
      likelihoodIidNormal( sim.df[,i], observ_solling_mat[,i_site], err_v[i] ) # make sure that here you compare the observation of correct site 
    }) %>% sum(.)
    
    if(is.nan(logpost) | is.na(logpost) | logpost == 0 ){logpost <- -Inf}
    
    logpost_all_sites <- logpost_all_sites + logpost
  }
  
  return( logpost_all_sites)
}

```

And then plug this function to the bayesian calibration.

I hope this is helpful.

Regards

**Comment by GakenouOF** [comment]
Thank you @trotsiuk, i will give this a go and see what happens. Thank you for your kind assistance.

**Comment by twest820** [comment]
> But, as described in the vignette text, the calibration run for **4e+06** iteration: `We will run 3 parallel chains, each on a separate computer node for 4e+06 iterations.`

Just to chip in here, I've been doing some numerical experiments on chain convergence rates with r3PG's 3-PGpjs 2.7 implementation as idle compute time comes available. Some observations from my setup, which may or may not transfer to other 3-PG calibration exercises:

- DEzs converges substantially more quickly, in a Gelman-Rubin multivariate R sense, than DE, DREAM, and DREAMzs. I've found it useful to run something like 100k to 500k iterations, log the Gelman-Rubin R, and continue to the next chunk of iterations. So far, increasing `nrChains` above 1 has been detrimental. I haven't as yet tried shifting any of the differential evolution MCMCs off their default number of correlated chains but it's worth noting DE defaults to three times the number of parameters worth of correlated chains. This is about `nParameters` more than the other options and yields correspondingly shorter chains. See discussion in  https://github.com/florianhartig/BayesianTools/issues/224.
- As one would likely expect, convergence rates decline as the number of parameters increases. I'm getting R < 1.1 within 100k DEzs iterations in contrived cases with single digit parameter counts. For a few tens of 3-PG parameters, my results are consistent with @trotsiuk's. Plan for chains with single digit millions of iterations. It looks to me like there are also important relationships between parameter significance to the model, the shapes of parameters' posteriors, and convergence rates. So there is some need to look at what chains are doing and take that back to the `BayesianSetup`. In practical cases I would speculate this isn't a hands off, give the chain to MCMC and let it turn the computational crank to get a model calibration thing.
- Memory allocations are volatile. R defaults to a memory limit matching the system's physical DRAM but there's 3-5GB of bounce every few seconds as a chain extends and sometimes that goes higher. There's additional complexity with R unloading parts of objects from memory and not I'm entirely convinced there's isn't some amount of memory leakage (or at least unnecessary memory accumulation) somewhere in the R-r3PG-BayesianTools system. For systems with DRAM constraints, I've found pushing up the default with `memory.limit()` to be necessary to prevent chaining failure on out of memory errors. Occasionally saving the chain (_e.g._ `mcmc_out`) and restarting R appears so far to be somewhat helpful, though I'm not sure this is necessary.
- I'm still working out upper bounds of what chain lengths fit in what amount of physical memory. At this point I can say there's no obvious trouble with 2M DEzs iterations using three correlated chains in 16GB DRAM, provided the system doesn't have too many other memory consumers. Beyond 3-4M iterations may be trickier as the nominal memory size of a 3 chain DEzs in R is 3.75 GB/1M iterations. 32+GB looks preferable if such a system is available.
- In the profiling I've done so far, enabling parallel chains (`createBayesianSetup(parallel = TRUE)`) has offered negligible benefit and may give a few percent slowdown. It'd be nice to have efficient 4-8+ thread parallelism instead of waiting on a single thread but I think this is a research topic.

**Comment by florianhartig** [comment]
Hi Todd, 

Regarding the **convergence**: when you note that "So far, increasing nrChains above 1 has been detrimental", your results actually indicate the opposite, i.e. that the Gelman Rubin diagnostics based on 1 MCMC is faulty and you have to set nrChains > 1, which I tried to explain in my response to your question in https://github.com/florianhartig/BayesianTools/issues/224. If your Gelman-Rubin diagnostics for the single MCMC were reliable, they should be similar when calculating Gelman Rubin within and between MCMCs. The fact that they get worse when calculating with more independent MCMCs shows that the internal chains of a single MCMC are not independent, and can therefore not be reliably used to calculate the Gelman Rubin diagnostics. As I said, this has been our repeated observation when running these MCMC samplers in practice, thus the recommendation in the help to always run several samplers to calculate the diagnostics. 

For that reason, **I absolute advice against comparing convergence in the DE sampler family based on a single MCMC run!**

Regarding memory allocation: I think we are mixing too many things here. Whether the r3PG Fortran code has a memory leak could easily be tested by repeatedly calling it outside BT in a simple loop. Regarding BT: I recently noticed via a question that there is the problem that even when implementing thinning, the z-matrix is increased for each iteration, which can create very large z matrices, see https://github.com/florianhartig/BayesianTools/issues/179

In any case, all these points seem to be different from what the OP wanted, so I would suggest to close this issue and open new issues for specific points, in particular 

- If there is a memory leak in 3PG, please open an issue for this with a reproducible example
- If there is a memory problem with the BT samplers, please open an issue on the BT issue tracker 

As I said, I don't think there is a problem with the calibration advice given in the vignette, running nrChains = 3 is preferable. @trotsiuk, maybe some additional comments in the light of this discussion could be added to clarify the questions that came up, e.g. a note that the DEzs has internal chains, but that convergence diagnostics based on several MCMCs are preferable. If you wanted to, I could do this as well.

**Comment by twest820** [comment]
> your results actually indicate the opposite, i.e. that the Gelman Rubin diagnostics based on 1 MCMC is faulty and you have to set nrChains > 1

Hi Florian, thank you! I'm afraid I'm finding this a confusing response which raises more questions, though. I'm just calling `gelmanDiagnostics()` in BayesianTools, likely the same as the OP as that's what the r3PG vignette illustrates. And I don't see an open issue for in the BayesianTools repo? Also, I haven't posted my observations of Gelman-Rubin R as a function of algorithm, number of iterations, `nrChain`, and so on. So it kinda seems you've made a conclusion about what I'm doing without the evidence upon which the conclusion is based? That doesn't mean the conclusion's incorrect, though.

Could you please clarify what's meant by "single MCMC"? Is a single MCMC a group/evolutionary population of correlated chains? A single chain within such a group? Multiple calls to `runMCMC()` which are merged later? For all the time I've spent with the BayesianTools source code so far, the internal data structures of the chains remain unclear to me. So I feel I'm lacking an ability to properly reconstruct the context you're speaking from.

This snippet below is what's currently in the r3PG vignette. Am I correct in thinking the `nrChains = 3` bit is what [Trotsiuk et al. 2020](https://doi.org/10.1111/2041-210X.13474) are referring to when indicating the use of three independent DEzs chains?
```R
mcmc_out <- runMCMC(
  bayesianSetup = mcmc_setup, 
  sampler = "DEzs",
  settings = list(iterations = 4e+03, nrChains = 3))
gelmanDiagnostics( mcmc_out )
```

> If your Gelman-Rubin diagnostics for the single MCMC were reliable, they should be similar when calculating Gelman Rubin within and between MCMCs. 

If a single MCMC is the group of correlated chains yielded from a call to `runMCMC()` with, say, DEzs and `nrChains = 1`, would that be a within MCMC comparison? And, if so, would a between MCMC comparison be comparing Gelman-Rubin R values from multiple `runMCMC()` calls using the same algorithm? Or does between MCMC mean between different algorithms, such as DEzs versus DREAM? If I have some DEzs chains that came from `nrChains = 3` is that between or within chains? Or both?

It also sounds like maybe there should be an issue open in BayesianTools for `gelmanDiagnostics()` being unreliable? I'm not seeing one, though.

> As I said, this has been our repeated observation when running these MCMC samplers in practice, thus the recommendation in the help to always run several samplers to calculate the diagnostics.

The [help for `gelmanDiagnostics()`](https://www.rdocumentation.org/packages/BayesianTools/versions/0.1.7/topics/gelmanDiagnostics) indicates concerns in this direction, which I think you're stating more strongly here, but the help is curiously specific about DEzs and I'm uncertain the current help text is really communicating the desired message. I also just checked back on the [BayesianTools vignette](https://github.com/florianhartig/BayesianTools/blob/master/BayesianTools/vignettes/BayesianTools.Rmd) and the `gelmanDiagnostics()` calls it makes are the same as what's currently in the r3PG vignette. However, those examples  are for Metropolis. Perhaps it would be good to also illustrate diagnostic calculations for one or more more of the DE algorithms on the BayesianTools side, possibly with annotations about choice of `nrChains` or other important parameters.

Something I'm also unclear on is the role of the autoburnin under `gelmanDiagnostics()`. `BayesianTools::gelmanDiagnostics()` calls `coda::gelman.diag()` with its default of `autoburnin = TRUE`. From the MCMC literature I've reviewed, it seems quite a few authors have done their chains in ways where autoburnin should result in consideration of only the parts of the correlated chains created under the default `nrChains = 1` which are "adequately independent-ish" for the resulting Gelman-Rubin to be deemed accurate enough. Your response makes me wonder if there are methodological concerns here which haven't adequately reflected in the set of MCMC publications I've happened to intersect with. If there's a specific paper users should be looking at that would be good to know. The current reference for `gelmanDiagnostics()` is ter Braak et al. 2008, which cites Gelman et al.'s 2004 book several times in emphasizing Gelman-Rubin R is valid for DEzs, but doesn't go beyond that.

> In any case, all these points seem to be different from what the OP wanted

Mmm, at the moment the central answer is at least some millions of iterations, per @trotsiuk. I checked back on Trotsiuk et al. 2020 and it says 24M iterations for _Picea abies_ and 17M for _Fagus sylvatica_ for R < 1.1. It's certainly my experience with these two packages and the computational hardware that's available around here that this number of iterations isn't something which just works. So it's probably useful to go beyond just saying crank up the iteration count.

In particular, I think there's relevant considerations about chain parallelism, multi-core use, DRAM requirements, and multi-machine considerations which haven't been discussed as yet in order for users to understand how to most effectively call into r3PG and BayesianTools to reach such iteration counts. For example, I was expecting `createBayesianSetup(parallel = TRUE)` with `runMCMC(nrChains = 3)` to cause three cores to go hot but instead only one core becomes active (several Rscript.exe processes are created alongside the original Rsession.exe, all but two of the Rscript.exes are cold and the remaining two plus the Rsession.exe add up to one core). If there is a way to get multiple cores doing independent DEzs or one of the other parallelized DE-MCMCs concurrently it seems valuable to have that illustrated somewhere in a vignette. A million DEzs+r3PG iterations takes a single core on most of the machines I have about 16 hours, implying two weeks for 20M iterations if DRAM or other constraints don't become limiting. So, for example, if the objective's to obtain three independent chains from `nrChains =3`, it's rather more convenient to run three cores concurrently for two weeks instead of one core for six weeks.

It seems to me an implication from the above is accuracy of Gelman-Rubin R may increase as `nrChains` increases. If so, wouldn't configurations of `nrChains = <number of cores>` or `nrChains = <number of hyperthreads>` and thus either `nCores` or `2*nCores` threads concurrently running independent chains be preferred for wall clock efficiency in a single machine case? (Not sure if there's a way to profile these options with the current packages, but whether one thread per core or two threads is most efficient probably depends on what the Fortran compiler decided to do with the innards of r3PG's loops.)

**Comment by florianhartig** [comment]
Hi Todd,

with the issues, I just meant please open a separate issue for each problem, else this gets terribly confusing for us to keep track of. This thread was originally about the question why @GakenouOF gets different plots than we do in the vignette. I think this problem is solved, so we should close. 

Other topics that have come up are a possible memory leak (if you think there is a memory leak please open a separate issue), and now we are mostly discussing issues of the BayesianTools documentation and convergence checks for the DE sampler family, which should be moved to the BayesianTools repo, as it has nothing to do with r3PG. 

I have provided some code that shows the Gelman Diagnostics problem here https://github.com/florianhartig/BayesianTools/issues/226 - if you have further concerns about the issue, please respond there. I maintain that I recommend to run several chains, backed up by he example that I provide, regardless of what may have been written in previous papers. 

Parallelization in BT is explained in https://cran.r-project.org/web/packages/BayesianTools/vignettes/BayesianTools.html#reference-on-creating-likelihoods and https://cran.r-project.org/web/packages/BayesianTools/vignettes/InterfacingAModel.html. Our experience with r3PG is that, for a single model run, within chain parallelization doesn't pay off against communication overhead, so the recommendation would be to simply run 3 independent R sessions with the 3 MCMCs and then merge them by hand. Alternatively, you can do this in the parallel environment by hand.
